src/main/scala/org/apache/sysml/compiler/lang/source/DML.scala

```diff
@@ -0,0 +1,473 @@
+               to the MLContext so that the actual dataframe reference will be passed with the name "ref"
+             */
+            case u.TermName("fromDataFrame") => args(0)
+          }
+        }
+
+        val printBuiltin = (target: D, sym: u.MethodSymbol, argss: Seq[Seq[D]], env: Environment) => {
+          val args = argss flatMap (args => args map (arg => arg(env)))
+
+          sym.name match {
+            case u.TermName("read") => {
+              s"""read(${args(0)}, format="${args(1).toLowerCase()}")"""
+            }
+
+            case u.TermName("write") => {
+              val format = args(2) match {
+                case "CSV" => """format="csv""""
+                case _ => throw new RuntimeException(s"Unsopported output format: ${args(2)}")
+              }
+              s"write(${args(0)}, ${args(1)}, $format)"
+            }
+
+            case u.TermName("variance") => s"var(${args.mkString(", ")})"
+
+            case u.TermName(fname) if fname == "MatrixOps" => args.head.toString
+
+            case u.TermName(fname) => s"$fname(${args.mkString(", ")})"
+
+            case _ =>
+              abort(s"Unsopported builtin call: $sym", sym.pos)
+          }
+        }
+
+        val isUnary = (sym: u.MethodSymbol) =>
+          sym.name.decodedName.toString.startsWith("unary_")
+
+        /**
+          * Convert Scala types to SystemML types
+          * @param arg the type argument
+          * @return DML type as string
+          */
+        def convertTypes(arg: u.Type): List[String] = arg match {
+          case tpe if tpe =:= u.typeOf[Double] => List("double")
+          case tpe if tpe =:= u.typeOf[Matrix] => List("matrix[double]")
+          case tpe if tpe <:< u.typeOf[Product] => {
+            val params = tpe.finalResultType.typeArgs
+            params.flatMap(convertTypes(_))
+          }
+          case tpe => abort(s"Unsupported return type $tpe. Supported types are: Matrix, Double")
+        }
+
+        // format a block to have 2 whitespace indentaion
+        val indent = (str: String) => {
+          val lines = str.split("\n")
+
+          if (lines.length > 1)
+            lines.map(x => s"  $x").mkString("\n")
+          else
+            s"  $str"
+        }
+
+        /**
+          *constructs a for loop by deconstructing foreach together with the range and the lambda */
+        val forLoop = (target: D, targs: Seq[u.Type], args: Seq[D], env: Environment) => {
+          val range = target(env)
+          val lambda = args.map(x => x(env)).head
+
+          val parts = lambda.split(" => ")
+          val idx = parts(0).drop(1).dropRight(1) // remove braces
+          // format the body with 2 spaces of indentation
+          val body = indent(parts(1))
+
+          s"""
+             |for ($idx in $range) {
+             |$body
+             |}
+            """.stripMargin.trim
+        }
+
+        val escape = (str: String) => str
```

Comment:
You should escape backslashes first, not last. Otherwise you'll replace newline with `\\n` instead of `\n`